textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/keywordmatch/KeywordMatcher.java

```diff
@@ -0,0 +1,180 @@
+     *           it won't match any document
+     *
+     *
+     *
+     */
+    @Override
+    public ITuple getNextTuple() throws DataFlowException {
+
+        List<IField> fieldList;
+        boolean foundFlag = false;
+        Set<String> setOfFoundTokens = new HashSet<>();
+        int positionIndex = 0; // Next position in the field to be checked.
+        int spanStartPosition; // Starting position of the matched query
+
+        try {
+            ITuple sourceTuple = sourceOperator.getNextTuple();
+            if(sourceTuple == null){
+                return null;
+            }
+            fieldList = sourceTuple.getFields();
+            spanList.clear();
+            if(!spanSchemaDefined){
+                Schema schema = sourceTuple.getSchema();
+                spanSchema = Utils.createSpanSchema(schema);
+                spanSchemaDefined = true;
+            }
+
+            for(int attributeIndex = 0; attributeIndex < attributeList.size(); attributeIndex++){
+                IField field = sourceTuple.getField(attributeList.get(attributeIndex).getFieldName());
+                String fieldValue = (String) (field).getValue();
+                String fieldName;
+                if(field instanceof StringField){
+                    //Keyword should match fieldValue entirely
+                    if(fieldValue.equalsIgnoreCase(query)){
+                        spanStartPosition = 0;
+                        positionIndex = query.length();
+                        fieldName = attributeList.get(attributeIndex).getFieldName();
+                        addSpanToSpanList(fieldName, spanStartPosition, positionIndex, query, fieldValue);
+                        foundFlag = true;
+                    }
+                }
+                else if(field instanceof TextField) {
+                    //Each element of Array of keywords is matched in tokenized TextField Value
+                    for(int iter = 0; iter < queryTokens.size(); iter++) {
+                        positionIndex = 0;
+                        String queryToken = queryTokens.get(iter);
+                        //Ex: For keyword lin it obtains pattern like /blin/b which matches keywords at boundary
+                        Pattern tokenPattern = tokenPatternList.get(iter);
+                        Matcher matcher = tokenPattern.matcher(fieldValue.toLowerCase());
+                        while (matcher.find(positionIndex) != false) {
+                            spanStartPosition = matcher.start();
+                            positionIndex = spanStartPosition + queryTokens.get(iter).length();
+                            String documentValue = fieldValue.substring(spanStartPosition, positionIndex);
+                            fieldName = attributeList.get(attributeIndex).getFieldName();
+                            addSpanToSpanList(fieldName, spanStartPosition, positionIndex, queryToken, documentValue);
+                            setOfFoundTokens.add(queryToken);
```

Comment:
The current "AND" implementation allows different keywords to appear in different fields to satisfy the "AND" condition.  I believe "AND" should mean all the query keywords should appear in a single field. In other words, we should go through the fields one by one and deal with them separately. For each field, if all the query tokens appear in the field, then we add its spans to the results. If one of the query tokens doesn't appear in the field, we ignore this field.

It's pretty straightforward to implement this logic.